src/hotspot/share/prims/jvmtiThreadState.inline.hpp

```diff
@@ -78,18 +78,22 @@ inline JvmtiThreadState* JvmtiThreadState::state_for_while_locked(JavaThread *th
 
   NoSafepointVerifier nsv;  // oop is safe to use.
 
-  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
-    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
-  }
-
   // In a case of unmounted virtual thread the thread can be null.
   JvmtiThreadState *state = thread == nullptr ? nullptr : thread->jvmti_thread_state();
 
   if (state == nullptr && thread != nullptr && thread->is_exiting()) {
     // Don't add a JvmtiThreadState to a thread that is exiting.
     return nullptr;
   }
-  if (state == nullptr || state->get_thread_oop() != thread_oop) {
+
+  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
+    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
+  }
+  // The state->get_thread_oop() may be null if the state is created during
+  // the allocation of the thread oop when a native thread is attaching. Make
+  // sure we don't create a new state for the JavaThread.
+  if (state == nullptr || (state->get_thread_oop() != nullptr &&
+                           state->get_thread_oop() != thread_oop)) {
     // Check if java_lang_Thread already has a link to the JvmtiThreadState.
     if (thread_oop != nullptr) {  // thread_oop can be null during early VMStart.
```

Comment:
This comment is another case of  `state->get_thread_oop()` being null. We should merge this comment with the new comment about attaching native thread.

Reply:
This also has caught by my eyes. :)
With the lines 99-101 in place the only case when `thread_oop` can be equal to `nullptr` is when `thread->threadObj() == nullptr`. My understanding is that can be for a detached thread only.
I would suggest to add an assert after the line 101:
  assert(thread_oop != nullptr, "sanity check");
Full testing with this assert should help to identify if it can be fired.
Then we can get rid of the check at the line 104.
The `JvmtiThreadState` constructor also allows for `thread_oop` to be `nullptr`.
Some cleanup will be needed to get rid of unneeded checks there as well.

Reply:
@sspitsyn For the above suggestions, it seems cleaner/safer to handle the clean-ups in a separate RFE with full testing including the vthread cases. There are additional comments in https://github.com/openjdk/jdk/pull/16642#issuecomment-1815379890 related to this as well. Those could be handled together and require through testing including the vthread support.

Reply:
Okay, thanks! I was about to suggest the same as it will simplify the process.
I've filed the following issue and assigned to myself:
[8320925](https://bugs.openjdk.org/browse/JDK-8320925 )`assert and cleanup in JvmtiThreadState::state_for_while_locked for thread_oop != nullptr`

Reply:
Thank you @sspitsyn, I'm interested in reviewing the followup change for [8320925](https://bugs.openjdk.org/browse/JDK-8320925) as well.